File: src/filter.ts
/**
 * filter.ts
 * Search and filter logic for article arrays. Pure functions only.
 * Given arrays in, returns filtered arrays out. No DOM. No localStorage.
 */

import { Article, Filter } from './models';

/**
 * Given an articles array and a Filter object, returns the filtered articles array.
 */
export function applyFilters(articles: Article[], filters: Filter): Article[] {
  let result = articles;
  if (filters.status) result = filterByDecision(result, filters.status);
  if (filters.tag)    result = filterByTag(result, filters.tag);
  if (filters.query)  result = searchArticles(result, filters.query);
  return result;
}

/**
 * Given an articles array and a decision string, returns articles matching that decision.
 */
export function filterByDecision(articles: Article[], decision: string): Article[] {
  return articles.filter(a => a.decision === decision);
}

/**
 * Given an articles array and a tag string, returns articles containing that tag.
 */
export function filterByTag(articles: Article[], tag: string): Article[] {
  const t = tag.trim().toLowerCase();
  return articles.filter(a => a.tags.some(x => x.toLowerCase() === t));
}

/**
 * Given an articles array and a query string, returns articles matching title or abstract.
 */
export function searchArticles(articles: Article[], query: string): Article[] {
  const q = query.trim().toLowerCase();
  if (!q) return articles;
  return articles.filter(
    a => a.title.toLowerCase().includes(q) || a.abstract.toLowerCase().includes(q),
  );
}

/**
 * Given an articles array, returns a sorted list of all unique tags in use.
 */
export function getAllTags(articles: Article[]): string[] {
  const tags = new Set<string>();
  articles.forEach(a => a.tags.forEach(t => tags.add(t)));
  return [...tags].sort();
}

/**
 * Given an Article and a tag string, returns the Article with that tag added if not already present.
 */
export function addTag(article: Article, tag: string): Article {
  const trimmed = tag.trim();
  if (!trimmed || article.tags.includes(trimmed)) return article;
  return { ...article, tags: [...article.tags, trimmed] };
}

/**
 * Given an Article and a tag string, returns the Article with that tag removed.
 */
export function removeTag(article: Article, tag: string): Article {
  return { ...article, tags: article.tags.filter(t => t !== tag) };
}

/**
 * Given an Article and a notes string, returns the Article with its notes field updated.
 */
export function updateNotes(article: Article, notes: string): Article {
  return { ...article, notes };
}

File: src/filter.py
"""Search and filter logic for article lists. All functions are pure with no side effects."""
from __future__ import annotations
from .models import Article, Filter


def apply_filters(articles: list[Article], filters: Filter) -> list[Article]:
    """Given an articles list and a Filter object, returns the filtered list."""
    result = articles

    if filters.status:
        result = filter_by_decision(result, filters.status)

    if filters.tag:
        result = filter_by_tag(result, filters.tag)

    if filters.query:
        result = search_articles(result, filters.query)

    return result


def filter_by_decision(articles: list[Article], decision: str) -> list[Article]:
    """Given an articles list and a decision string, returns articles matching that decision."""
    return [a for a in articles if a.decision == decision]


def filter_by_tag(articles: list[Article], tag: str) -> list[Article]:
    """Given an articles list and a tag string, returns articles containing that exact tag."""
    tag_lower = tag.strip().lower()
    return [a for a in articles if any(t.lower() == tag_lower for t in a.tags)]


def search_articles(articles: list[Article], query: str) -> list[Article]:
    """Given an articles list and a query string, returns articles matching title or abstract."""
    q = query.strip().lower()
    if not q:
        return articles
    return [
        a for a in articles
        if q in a.title.lower() or q in a.abstract.lower()
    ]


def get_all_tags(articles: list[Article]) -> list[str]:
    """Given an articles list, returns a sorted list of all unique tags currently in use."""
    tags: set[str] = set()
    for article in articles:
        tags.update(article.tags)
    return sorted(tags)


def add_tag(article: Article, tag: str) -> Article:
    """Given an article and a tag string, returns the article with that tag added if not already present."""
    tag = tag.strip()
    if tag and tag not in article.tags:
        return article.model_copy(update={"tags": article.tags + [tag]})
    return article


def remove_tag(article: Article, tag: str) -> Article:
    """Given an article and a tag string, returns the article with that tag removed."""
    return article.model_copy(update={"tags": [t for t in article.tags if t != tag]})


def update_notes(article: Article, notes: str) -> Article:
    """Given an article and a notes string, returns the article with its notes field updated."""
    return article.model_copy(update={"notes": notes})
